File: lesson-08/js/RubberArrayRun.js
class RubberArray {
    static DEFAULT_CAPACITY = 2;

    constructor() {
        this.array = new Array(RubberArray.DEFAULT_CAPACITY);
        this.size = 0;
    }

    append(element) {
        this.ensureCapacity(this.size + 1);
        this.array[this.size++] = element;
    }

    appendRange(elements) {
        const requiredCapacity = this.size + elements.length;
        this.ensureCapacity(requiredCapacity);
        for (let i = 0; i < elements.length; i++) {
            this.array[this.size++] = elements[i];
        }
    }

    remove(element) {
        const index = this.array.indexOf(element);
        if (index !== -1) {
            this.removeAt(index);
        }
    }

    removeAt(position) {
        if (position < 0 || position >= this.size) {
            throw new Error("Неверный индекс");
        }
        for (let i = position; i < this.size - 1; i++) {
            this.array[i] = this.array[i + 1];
        }
        this.size--;
    }

    get(index) {
        if (index < 0 || index >= this.size) {
            throw new Error("Неверный индекс");
        }
        return this.array[index];
    }

    size() {
        return this.size;
    }

    contains(element) {
        return this.array.indexOf(element) !== -1;
    }

    ensureCapacity(minCapacity) {
        const currentCapacity = this.array.length;
        if (minCapacity > currentCapacity) {
            const newCapacity = Math.max(minCapacity, currentCapacity * 2);
            this.array = this.copyArray(newCapacity);
        }
    }

    copyArray(newCapacity) {
        const newArray = new Array(newCapacity);
        for (let i = 0; i < this.size; i++) {
            newArray[i] = this.array[i];
        }
        return newArray;
    }

    toString() {
        return `[${this.array.slice(0, this.size).join(" ")}]`;
    }
}

const main = () => {
    const ra = new RubberArray();
    for (let i = 1; i <= 5; i++) {
        ra.append(i * 10);
    }
    console.log(ra.toString()); // [10 20 30 40 50]
    ra.remove(20);
    console.log(ra.toString()); // [10 30 40 50]

}

main();

File: lesson-08/java/RubberArrayRun.java

public class RubberArrayRun {
    public static void main(String[] args) {
        RubberArray ra = new RubberArray();
        for (int i = 1; i <= 5; i++) {
            ra.append(i * 10);
        }
        System.out.println(ra); // [10 20 30 40 50]
        ra.remove(20);
        System.out.println(ra); // [10 30 40 50]

    }
}

class RubberArray {
    private int[] array;
    private int size;
    private static final int DEFAULT_CAPACITY = 2;

    public RubberArray() {
        this.array = new int[DEFAULT_CAPACITY];
        this.size = 0;
    }

    public void append(int element) {
        ensureCapacity();
        array[size++] = element;
    }

    public void appendRange(int[] elements) {
        ensureCapacity(size + elements.length);
        System.arraycopy(elements, 0, array, size, elements.length);
        size += elements.length;
    }

    public void remove(int element) {
        for (int i = 0; i < size; i++) {
            if (array[i] == element) {
                removeAt(i);
                break;
            }
        }
    }

    public void removeAt(int position) {
        if (position < 0 || position >= size) {
            throw new IndexOutOfBoundsException("Invalid index");
        }
        System.arraycopy(array, position + 1, array, position, size - position - 1);
        size--;
    }

    public int get(int index) {
        if (index < 0 || index >= size) {
            throw new IndexOutOfBoundsException("Invalid index");
        }
        return array[index];
    }

    public int size() {
        return size;
    }

    public boolean contains(int element) {
        for (int i = 0; i < size; i++) {
            if (array[i] == element) {
                return true;
            }
        }
        return false;
    }

    private void ensureCapacity() {
        if (size == array.length) {
            int newCapacity = array.length * 2;
            array = copyArray(array, newCapacity);
        }
    }

    private void ensureCapacity(int minCapacity) {
        if (minCapacity > array.length) {
            int newCapacity = Math.max(minCapacity, array.length * 2);
            array = copyArray(array, newCapacity);
        }
    }

    private int[] copyArray(int[] original, int newCapacity) {
        int[] newArray = new int[newCapacity];
        for (int i = 0; i < original.length; i++) {
            newArray[i] = original[i];
        }
        return newArray;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("[");
        for (int i = 0; i < size; i++) {
            sb.append(array[i]);
            if (i < size - 1) {
                sb.append(" ");
            }
        }
        sb.append("]");
        return sb.toString();
    }
}
